cli/lsp/documents.rs

```diff
@@ -664,6 +671,7 @@ impl SpecifierResolver {
 #[derive(Debug, Default)]
 struct FileSystemDocuments {
   docs: HashMap<ModuleSpecifier, Document>,
+  metadata: HashMap<ModuleSpecifier, Arc<HashMap<MetadataKey, String>>>,
```

Comment:
I think this specifier metadata and #12864 are not really document related, but rather have to do with specifiers. Also, `FileSystemDocuments` contains only non-open documents.

Maybe we should move this out to a separate struct on its own and then I think we'd only need to use that struct in diagnostics.rs? (So we could remove most of the code in this file)

Reply:
I disagree... they are very much "FileSystemDocuments" as they are what is stored in the cache... while we don't load 100% of files in the cache as document nor do 100% of documents that are loaded get their meta data loaded, they both are representations of information in the cache and have a coupled shared state.

Documents open in the editor for editing can't have metadata, only `FileSystemDocuments` can (and only those in the cache).

Reply:
I think they should share the same underlying cache, but we should separate this out from `FileSystemDocuments`. The purpose of `FileSystemDocuments` is to prevent having multiple copies of the same `Document` loaded into memory.

I would also say this shouldn't be on the interface for `Documents` either. We're wanting to know what the metadata is for certain specifiers and not documents. For example, there could be two specifiers that point at the same document and only one of those specifiers would produce a certain diagnostic. Again, they would share the same underlying cache.

Maybe I'm misunderstanding the purpose of this though.